torch/_prims/__init__.py

```diff
@@ -0,0 +1,992 @@
+    "reciprocal",
+    "round",
+    "sign",
+    "sin",
+    "sinh",
+    "sqrt",
+    "square",
+    "tan",
+    #
+    # Elementwise binary prims
+    #
+    "add",
+    "atan2",
+    "bitwise_and",
+    "bitwise_not",
+    "bitwise_or",
+    "bitwise_xor",
+    # 'complex',  # needs custom meta
+    "div",
+    "eq",
+    "ge",
+    "gt",
+    "le",
+    "lt",
+    "max",
+    "min",
+    "mul",
+    "ne",
+    "nextafter",
+    "pow",
+    "rsqrt",
+    "shift_left",
+    "shift_right_arithmetic",
+    "shift_right_logical",  # not implemented
+    #
+    # View prims
+    #
+    "broadcast_in_dim",
+    "collapse_view",
+    "split_dim",
+    "squeeze",
+    #
+    # Shape prims
+    #
+    "collapse",
+    "concatenate",
+    "reshape",
+    #
+    # Conditional prims
+    #
+    "select",
+    #
+    # Data conversion and movement prims
+    #
+    "convert_element_type",
+    "device_put",
+    #
+    # Inplace prims
+    #
+    "copy_to",
+    "resize",
+]
+
+#
+# Common datastructures and helpers
+#
+
+# Describes the return type of the primitive:
+#
+#   - NEW, a new tensor is created
+#   - VIEW, a view of an input tensor is returned
+#   - INPLACE, one or more input tensors is modified
+#
+# these descriptors are mututally exclusive and exhaustive.
+class RETURN_TYPE(Enum):
+    NEW = (0,)
+    VIEW = (1,)
+    INPLACE = (2,)
+
+
+def _make_prim(
+    *, meta: Callable, impl_aten: Callable, return_type: RETURN_TYPE, doc: str
+):
+    """
+    Creates a primitive operation.
+
+    """
+
+    def _prim(*args, **kwargs):
+        # TODO: refactor this to not incur such a latency hit
+        return_meta = False
+        for arg in args:
+            if isinstance(arg, TensorMeta):
+                return_meta = True
+        for v in kwargs.values():
+            if isinstance(v, TensorMeta):
+                return_meta = True
+        meta_result = meta(*args, **kwargs)
+
+        if return_meta:
+            return meta_result
+
+        return impl_aten(*args, **kwargs)
+
+    _prim.__doc__ = doc
+    _prim.meta = meta
+    _prim.return_type = return_type
+
+    return _prim
+
+
+def _elementwise_meta(*args):
+    '''
+    Meta function for elementwise operations.
+
+    Stride logic is currently incorrect.
+    '''
+
+    assert len(args) > 0
+
+    utils.check_same_device(*args, allow_scalars=True)
+    utils.check_same_shape(*args)
+    utils.check_same_dtype(*args)
+
+    strides = None
+    tensor = None
+    number = None
+    for arg in args:
+        if isinstance(arg, TensorLike):
+            if strides is None:
+                strides = arg.stride()
+
+            if tensor is None:
+                tensor = arg
+
+            if arg.stride() != strides:
+                return TensorMeta(
+                    arg, strides=utils.make_contiguous_strides_for(arg.shape)
+                )
+        elif isinstance(arg, Number):
+            if number is None:
+                number = arg
+
+    # TODO: fix strides
```

Comment:
the hard part 😱

Reply:
Not sure what the plan is, actually. Maybe you don't want prims to inherit the insane stride preservation logic that TI does?

Reply:
What's insane about it :-) ? We've been talking about whether prims should guarantee strides, and I'm openminded that they shouldn't, but then, it would be nice if they did? Anyway, we are punting on it for now.

Reply:
Insane in a good way :) I have been asking forever for a clear reference implementation of the logic in Python haha

Reply:
Just to see if I understand this. this "insane preservation logic" is Ed is talking about that algorithm that, given a some strides (perhaps not contiguous in memory) it chooses the best contiguous (as in contiguous in memory) strides for the output that provide the best way to iterate over them?

Reply:
yup